feat: paraswap permit

defispartan opened the pull request:
## General Changes

- Add support to approve repay with collateral and collateral swap transactions with gasless signature in place of approval tx. 

## Developer Notes

Signature request is generated in new helper function in `poolSlice`. This should be moved to @aave/contract-helpers, bundled with approval tx building, and re-used for pool permit functions, staking, and migration tool as started [here](https://github.com/aave/aave-utilities/pull/458)

Refactoring this before the migration PR will cause large conflicts though, so this approach is temporary.

---

## Author Checklist

Please ensure you, the author, have gone through this checklist to ensure there is an efficient workflow for the reviewers.

- [x]  The base branch is set to `main`
- [x]  The title is using [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) formatting
- [ ]  The Github issue has been linked to the PR in the Development section
- [x]  The General Changes section has been filled out
- [x]  Developer Notes have been added (optional)

**If the PR is ready for review:**

- [x]  The PR is in `Open` state and not in `Draft` mode
- [ ]  The `Ready for Dev Review` label has been added

## Reviewer Checklist

Please ensure you, as the reviewer(s), have gone through this checklist to ensure that the code changes are ready to ship safely and to help mitigate any downstream issues that may occur.

- [ ]  End-to-end tests are passing without any errors
- [ ]  Code style generally follows existing patterns
- [ ]  Code changes do not significantly increase the application bundle size
- [ ]  If there are new 3rd-party packages, they do not introduce potential security threats
- [ ]  If there are new environment variables being added, they have been added to the `.env.example` file as well as the pertinant `.github/actions/*` files
- [ ]  There are no CI changes, or they have been approved by the DevOps and Engineering team(s)
- [ ]  Code changes have been quality checked in the ephemeral URL
- [ ]  QA verification has been completed
- [ ]  There are two or more approvals from the core team
- [ ]  Squash and merge has been checked

grothem commented:
lint T-1876

MareskoY commented:
@defispartan  all actions related to paraswap and usdt on any market failing (for repay as collateral and swap )

iamanastasia commented:
Not sure if that's critical, but might be confusing in UI — error message stays on the screen, even when I'm trying to do a new action.

Path:
1. Decided to swap collateral and opened modal window.
2. Selected asset I want to swap to.
3. Clicked APPROVE (via signed message).
4. Clicked CANCEL in the wallet because I changed my mind.
5. Got an error message 'You canceled the transaction', which is fine.
6. Changed asset in 'Swap to'.
7. And I still see the old error message about 'You canceled the transaction' even though all other details were updated.

Shouldn we hide old errors since I'm trying to perform a new action?

<img width="415" alt="Screenshot 2023-01-10 at 15 20 20" src="https://user-images.githubusercontent.com/92739089/211537903-53dc6c7d-2d10-42c0-a095-6c13945d8418.png">

defispartan commented:
> Not sure if that's critical, but might be confusing in UI — error message stays on the screen, even when I'm trying to do a new action.
> 
> Path:
> 
> 1. Decided to swap collateral and opened modal window.
> 2. Selected asset I want to swap to.
> 3. Clicked APPROVE (via signed message).
> 4. Clicked CANCEL in the wallet because I changed my mind.
> 5. Got an error message 'You canceled the transaction', which is fine.
> 6. Changed asset in 'Swap to'.
> 7. And I still see the old error message 'You canceled the transaction' even though all other details were updated.
> 
> Should we hide old errors since I'm trying to perform a new action?
> 
> <img alt="Screenshot 2023-01-10 at 15 20 20" width="415" src="https://user-images.githubusercontent.com/92739089/211537903-53dc6c7d-2d10-42c0-a095-6c13945d8418.png">

Pushed an update, the error and approval state should be correctly updated on any changes to the selected asset or amount

iamanastasia commented:
@defispartan Looks great! thanks

MareskoY commented:
there is a problem with swap action for USDT on ETH and avalanche market, only on fork

it's every time failing with
```
cannot estimate gas; transaction may fail or may require manual gas limit (error={"reason":"processing response error","code":"SERVER_ERROR","body":"{\"id\":152,\"jsonrpc\":\"2.0\",\"error\":{\"code\":-32000,\"message\":\"execution reverted\",\"data\":{\"error_id\":\"fbb18036-3e4e-40ef-ab0e-dea6cbd4b11a\"}}}","error":{"code":-32000,"data":{"error_id":"fbb18036-3e4e-40ef-ab0e-dea6cbd4b11a"}},"requestBody":"{\"method\":\"eth_estimateGas\",\"params\":[{\"value\":\"0x0\",\"from\":\"0x9effcf18637ad365a5240b1b2db6ca7b46a5c98a\",\"to\":\"0x80aca0c645fedabaa20fd2bf0daf57885a309fe6\",\"data\":\"0x4db9dc97000000000000000000000000c02aaa39b223fe8d0a0e5c4f27ead9083c756cc2000000000000000000000000dac17f958d2ee523a2206206994597c13d831ec7000000000000000000000000000000000000000000000000001a9d3facc3d3500000000000000000000000000000000000000000000000000000000000989680000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000180000000000000000000000000000000000000000000000000001cfc002fec71d90000000000000000000000000000000000000000000000000000000063beb003000000000000000000000000000000000000000000000000000000000000001b4daaed2b88ecdf25bae8cd7662d3902b51a1e2e66cf159c083ebbc638bc6d63743061e5f10ba587bf3a0fa5a7b0e9cc49185254929ce41ec1e986388cb48c71c00000000000000000000000000000000000000000000000000000000000001600000000000000000000000000000000000000000000000000000000000000040000000000000000000000000def171fe48cf0115b1d80b88dc8eab59176fee5700000000000000000000000000000000000000000000000000000000000000e4b2f1e6db000000000000000000000000c02aaa39b223fe8d0a0e5c4f27ead9083c756cc2000000000000000000000000000000000000000000000000001a9d3facc3d3500000000000000000000000000000000000000000000000000000000000989680000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000a00000000000000000000000000000000000000000000000000000000000000001000000000000000000004de4703b120f15ab77b986a24c6f9262364d02f9432f00000000000000000000000000000000000000000000000000000000\"}],\"id\":152,\"jsonrpc\":\"2.0\"}","requestMethod":"POST","url":"https://rpc.tenderly.co/fork/ea6b30db-9e93-4c2d-952b-6cb71dde8ac6"}, method="estimateGas", transaction={"from":"0x9efFCf18637ad365A5240B1b2Db6CA7B46a5C98a","to":"0x80Aca0C645fEdABaa20fd2Bf0Daf57885A309FE6","value":{"type":"BigNumber","hex":"0x00"},"data":"0x4db9dc97000000000000000000000000c02aaa39b223fe8d0a0e5c4f27ead9083c756cc2000000000000000000000000dac17f958d2ee523a2206206994597c13d831ec7000000000000000000000000000000000000000000000000001a9d3facc3d3500000000000000000000000000000000000000000000000000000000000989680000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000180000000000000000000000000000000000000000000000000001cfc002fec71d90000000000000000000000000000000000000000000000000000000063beb003000000000000000000000000000000000000000000000000000000000000001b4daaed2b88ecdf25bae8cd7662d3902b51a1e2e66cf159c083ebbc638bc6d63743061e5f10ba587bf3a0fa5a7b0e9cc49185254929ce41ec1e986388cb48c71c00000000000000000000000000000000000000000000000000000000000001600000000000000000000000000000000000000000000000000000000000000040000000000000000000000000def171fe48cf0115b1d80b88dc8eab59176fee5700000000000000000000000000000000000000000000000000000000000000e4b2f1e6db000000000000000000000000c02aaa39b223fe8d0a0e5c4f27ead9083c756cc2000000000000000000000000000000000000000000000000001a9d3facc3d3500000000000000000000000000000000000000000000000000000000000989680000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000a00000000000000000000000000000000000000000000000000000000000000001000000000000000000004de4703b120f15ab77b986a24c6f9262364d02f9432f00000000000000000000000000000000000000000000000000000000","accessList":null}, code=UNPREDICTABLE_GAS_LIMIT, version=providers/5.5.3)
```
by this reason Ci failing, at same time on mainnet it's work fine

foodaka commented on src/components/transactions/Warnings/ParaswapErrorDisplay.tsx:
```diff
@@ -0,0 +1,31 @@
+import { GasEstimationError } from '../FlowCommons/GasEstimationError';
+
+const USER_DENIED_SIGNATURE = 'MetaMask Message Signature: User denied message signature.';
+const USER_DENIED_TRANSACTION = 'MetaMask Message Signature: User denied message signature.';
```
should this be wrapped in a trans?

defispartan replied:
It's not displayed so it doesn't need a Trans. These are used to check the error message returned by MetaMask to decide if the error is was from the user denying the action or gas estimation
